lib/Reader/readerir.cpp

```diff
@@ -5994,8 +5994,35 @@ void GenIR::switchOpcode(IRNode *Opr) {
   BasicBlock *CurrBlock = LLVMBuilder->GetInsertBlock();
   TerminatorInst *TermInst = CurrBlock->getTerminator();
   SwitchInst *SwitchInstruction = cast<SwitchInst>(TermInst);
-
   SwitchInstruction->setCondition(Opr);
+
+  Type *OprType = Opr->getType();
+  LLVMContext &LLVMContext = *this->JitContext->LLVMContext;
+  Type *Int32Ty = Type::getInt32Ty(LLVMContext);
+  if (OprType != Int32Ty) {
+    // When we made the case values we assumed int32 as the type for the
+    // values. But now we have an operand of a different type. LLVM
+    // requires the values and the switch value to have the same type.
+    // So change the type of the values to be the type of the selector.
+    IntegerType *OprIntType = cast<IntegerType>(OprType);
+
+    // If a small selector type was used, e.g. i8, the type might not
+    // be big enough to hold all the case values. But the value for the
+    // selector is popped off the evaluation stack, so it should be at
+    // least 32 bits. So check that.
+    unsigned OprBitWidth = OprIntType->getBitWidth();
+    assert((OprBitWidth > 32) && "Selector bit width is less then 32");
+
+    // Create new values of the correct type.
+    int NumCases = SwitchInstruction->getNumCases();
+    SwitchInst::CaseIt Case = SwitchInstruction->case_begin();
+    for (size_t I = 0; I < NumCases; I++) {
+      ConstantInt *Value = ConstantInt::get(
+          *JitContext->LLVMContext, APInt(OprBitWidth, (uint64_t)I, false));
```

Comment:
Sorry, I didn't mean this should assert something about the original type, I meant it should assert something about the original value.  Your example in e-mail was:
```
  switch i64 %23, label %24 [
    i32 6, label %109
    i32 1, label %26
    i32 2, label %41
    i32 3, label %48
    i32 4, label %52
    i32 5, label %64
  ]
```
which, as written, this code would extend to
```
  switch i64 %23, label %24 [
    i64 1, label %109
    i64 2, label %26
    i64 3, label %41
    i64 4, label %48
    i64 5, label %52
    i64 6, label %64
  ]
```
which would be incorrect.  I've looked at the code and I think it's true that given the way we generate these switches the value of the _n_ th case will always be _n_, but then again I don't know where the switch in the e-mail got permuted, an assertion would make sure no permuting has happened (or you could compute the new value based on the old value rather than based on the `for` loop iterator).